src/librustc_mir/transform/promote_consts.rs

```diff
@@ -293,17 +295,27 @@ impl<'a, 'tcx> Promoter<'a, 'tcx> {
         new_temp
     }
 
-    fn promote_candidate(mut self, candidate: Candidate) {
+    fn promote_candidate(
+        mut self,
+        def_id: DefId,
+        candidate: Candidate,
+        next_promoted_id: usize,
+    ) -> Option<Body<'tcx>> {
         let mut operand = {
             let promoted = &mut self.promoted;
-            let promoted_id = Promoted::new(self.source.promoted.len());
-            let mut promoted_place = |ty, span| {
+            let promoted_id = Promoted::new(next_promoted_id);
+            let mut promoted_place = |ty, tcx, span| {
```

Comment:
you don't need the `tcx` as an argument. Just do a `let tcx = self.tcx;` before creating the closure. This way you don't have to pass a `tcx` argument